File: time/src/main/java/com/tr/hsyn/time/duration/Duration.java
package com.tr.hsyn.time.duration;


import com.tr.hsyn.random.Randoom;
import com.tr.hsyn.scaler.Generatable;
import com.tr.hsyn.scaler.Generation;
import com.tr.hsyn.time.Time;
import com.tr.hsyn.time.Unit;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


/**
 * Allows specifying a <b><i>duration<i></b>.<br>
 * Durations are shown in specific units ({@link Unit}).<br><br>
 * <pre>
 *    var duration = Duration.of(Unit.MINUTE, 1981); // 1981 minutes duration
 *    //var duration = Duration.ofMinute(1981);// same as above
 * </pre>
 * <p>
 * There are no restrictions when specifying the duration,
 * the that minute is saved as it is.
 * So it is not treated as time, because if it were treated as time,
 * the minute value could be at most {@code 59}.<br>
 * However,
 * a <b>duration</b> can be converted to a time by calling {@link #toDurationGroup()}<br><br>
 *
 * <pre>
 * var duration = Duration.ofMinute(1981);
 * System.out.println(duration.toTimeDuration());
 * // [Duration{type=day, value=1}, Duration{type=hour, value=9}, Duration{type=minute, value=1}]
 * </pre>
 *
 * @see Unit
 * @see DurationGroup
 */
public interface Duration extends Comparable<Duration>, Generatable<Duration> {
	
	@NotNull
	String toString(@NotNull String formatted);
	
	/**
	 * @return the unit of this {@linkplain Duration} object
	 */
	@NotNull Unit getUnit();
	
	/**
	 * @return duration value of this {@linkplain Duration} object
	 */
	long getValue();
	
	@Override
	default @NotNull Generation<Duration> getGeneration(@NotNull Duration start, @NotNull Duration end, @NotNull Duration step) {
		
		return new DurationGeneration(start, end, step);
	}
	
	/**
	 * Determines whether this unit of the duration is equal to given unit of the duration.
	 *
	 * @param duration the duration
	 * @return {@code true} if this unit of the duration is equal to given unit of the duration.
	 */
	default boolean equalsByUnit(@NotNull Duration duration) {
		
		return getUnit().equals(duration.getUnit());
	}
	
	/**
	 * @return {@code true} if duration value is zero, {@code false} otherwise
	 */
	default boolean isZero() {
		
		return getValue() == 0L;
	}
	
	/**
	 * Adds {@code value} to the {@link Duration}.
	 * İf the given value is negative, made subtract.
	 *
	 * <pre>
	 *     var duration = Duration.of(Unit.MINUTE, 5);
	 *     var duration2 = Duration.ofMinute(-1);
	 *     var duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
	 *     var duration4 = duration.plus(-1L); // Duration{type=minute, value=4}
	 * </pre>
	 *
	 * @param value Amount of the time
	 * @return new {@link Duration}
	 */
	@NotNull
	default Duration plus(long value) {
		
		return of(getUnit(), getValue() + value);
	}
	
	/**
	 * Returns a new {@link Duration} with the given value.
	 * The unit is not changed.
	 *
	 * @param value Amount of the time
	 * @return new {@link Duration}
	 */
	default Duration withValue(long value) {
		
		return of(getUnit(), value);
	}
	
	/**
	 * @param other duration
	 * @return {@code true} if this {@link Duration} is less than other
	 */
	default boolean lessThan(@NotNull Duration other) {
		
		return this.compareTo(other) < 0;
	}
	
	/**
	 * @param other duration
	 * @return {@code true} if this {@link Duration} is less than other or equal
	 */
	default boolean lessThanOrEqual(@NotNull Duration other) {
		
		return this.compareTo(other) <= 0;
	}
	
	/**
	 * @param other duration
	 * @return {@code true} if this {@link Duration} is greater than other
	 */
	default boolean greaterThan(@NotNull Duration other) {
		
		return this.compareTo(other) > 0;
	}
	
	/**
	 * @param other duration
	 * @return {@code true} if this {@link Duration} is greater than other or equal
	 */
	default boolean greaterThanOrEqual(@NotNull Duration other) {
		
		return this.compareTo(other) >= 0;
	}
	
	/**
	 * Returns exactly how long the duration is as date format.
	 *
	 * <pre>
	 * var duration = Duration.ofMinute(1981);
	 * System.out.println(duration.toTimeDuration());
	 * // [Duration{type=day, value=1}, Duration{type=hour, value=9}, Duration{type=minute, value=1}]
	 * </pre>
	 *
	 * @return {@link DurationGroup}
	 */
	@NotNull
	default DurationGroup toDurationGroup() {
		
		return Time.toDuration(toMilliseconds());
	}
	
	/**
	 * @return as milliseconds of this {@linkplain Duration}
	 */
	default long toMilliseconds() {
		
		long value = getValue();
		
		//@off
		switch (getUnit()) {
			case MILLISECOND : return value;
			case SECOND :      return value * 1000;
			case MINUTE :      return value * 60000;
			case HOUR  :       return value * 3600000;
			case DAY :         return value * 86400000;
			case MONTH  :      return value * 259200000;
			case YEAR :        return value * 36500000;
		}//@on
		
		return 0;
	}
	
	/**
	 * Adds {@code Duration} to this {@link Duration}.
	 * İf the given duration is negative, made subtract.
	 * İf the given {@linkplain Duration} has a different {@linkplain Unit},
	 * it is converted to this {@linkplain Duration} unit.
	 * So returned {@link Duration} unit is same as this {@linkplain Duration} unit.
	 *
	 * <pre>
	 *     var duration = Duration.of(Unit.MINUTE, 5);
	 *     var duration2 = Duration.ofMinute(-1);
	 *     var duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
	 *     var duration4 = duration.plus(-1L); // Duration{type=minute, value=4}
	 * </pre>
	 *
	 * @param other Other {@linkplain Duration} object to add
	 * @return new {@link Duration}
	 */
	@NotNull
	default Duration plus(@NotNull Duration other) {
		
		if (!getUnit().equals(other.getUnit())) {
			
			Duration d = other.getAs(getUnit());
			return new DurationImp(getUnit(), getValue() + d.getValue());
		}
		
		return new DurationImp(getUnit(), getValue() + other.getValue());
	}
	
	/**
	 * Converts the duration to the given unit
	 *
	 * @param unit the unit
	 * @return {@link Duration}
	 */
	@NotNull
	default Duration getAs(@NotNull Unit unit) {
		
		long value = toMilliseconds();
		//@off
		switch (unit) {
			case MILLISECOND : return new DurationImp(unit, value);
			case SECOND      : return new DurationImp(unit, value / 1000);
			case MINUTE      : return new DurationImp(unit, value / 60000);
			case HOUR        : return new DurationImp(unit, value / 3600000);
			case DAY         : return new DurationImp(unit, value / 86400000);
			case MONTH       : return new DurationImp(unit, value / 259200000);
			case YEAR        : return new DurationImp(unit, value / 36500000);
		}//@on
		
		throw new IllegalArgumentException("Unknown unit: " + unit);
	}
	
	/**
	 * Creates a new Duration.
	 *
	 * @param unit  Unit of time
	 * @param value Amount of time
	 * @return Duration
	 */
	@NotNull
	static Duration of(@NotNull Unit unit, long value) {
		
		return new DurationImp(unit, value);
	}
	
	static void main(String[] args) {
		
		@NotNull Duration duration  = Duration.of(Unit.MINUTE, 5);
		@NotNull Duration duration2 = Duration.ofMinute(-1);
		Duration          duration3 = duration.plus(duration2); // Duration{type=minute, value=4}
		System.out.println(duration3.toString());
	}
	
	/**
	 * Creates a new Duration with given minutes.
	 *
	 * @param minutes Number of minutes
	 */
	@NotNull
	static Duration ofMinute(long minutes) {
		
		return of(Unit.MINUTE, minutes);
	}
	
	@NotNull
	static DurationGroup of(@NotNull List<Duration> durations) {
		
		return DurationGroup.of(durations);
	}
	
	@NotNull
	static DurationGroup of(@NotNull Duration... durations) {
		
		return DurationGroup.of(new ArrayList<>(Arrays.asList(durations)));
	}
	
	/**
	 * Creates a new Duration with value of zero.
	 *
	 * @param unit Unit of time
	 */
	@NotNull
	static Duration ofZero(@NotNull Unit unit) {return new DurationImp(unit, 0L);}
	
	/**
	 * Creates a new Duration with given years.
	 *
	 * @param years Number of years
	 */
	@NotNull
	static Duration ofYear(long years) {return of(Unit.YEAR, years);}
	
	/**
	 * Creates a new Duration with given months.
	 *
	 * @param months Number of months
	 */
	@NotNull
	static Duration ofMonth(long months) {
		
		return of(Unit.MONTH, months);
	}
	
	/**
	 * Creates a new Duration with given days.
	 *
	 * @param days Number of days
	 */
	@NotNull
	static Duration ofDay(long days) {return of(Unit.DAY, days);}
	
	/**
	 * Creates a new Duration with given hours.
	 *
	 * @param hours Number of hours
	 */
	@NotNull
	static Duration ofHour(long hours) {
		
		return of(Unit.HOUR, hours);
	}
	
	/**
	 * Creates a new Duration with given seconds.
	 *
	 * @param seconds Number of seconds
	 */
	@NotNull
	static Duration ofSecond(long seconds) {
		
		return of(Unit.SECOND, seconds);
	}
	
	/**
	 * Creates a new Duration with given milliseconds.
	 *
	 * @param milliseconds Number of milliseconds
	 **/
	@NotNull
	static Duration ofMillisecond(long milliseconds) {
		
		return of(Unit.MILLISECOND, milliseconds);
	}
	
	interface Random {
		
		long MAX_VALUE = 100L;
		
		@NotNull
		static Duration getNext() {
			
			return new DurationImp(Unit.values()[Randoom.getInt(Unit.values().length)], Randoom.getLong(MAX_VALUE));
		}
	}
}

File: timek/src/main/java/tr/xyz/timek/Duration.kt
package tr.xyz.timek

import java.util.*
import kotlin.reflect.KProperty

/**
 * Represents a duration.
 *
 * Duration is an amount of time with certain unit (year, month, day, hour, minute, second, millisecond).
 * And this duration **is not** a time duration.
 * So, the time units have no limit.
 * For example, can be defined a duration of a month with value `2_000`.
 *
 * ```
 *
 *    val _duration = Duration months 2000
 *    println(_duration) // 2000 months
 * ```
 *
 * And a duration can be converted to any unit.
 *
 * ```
 *
 *    val _duration = Duration months 2000 to TimeUnit.YEAR
 *    println(_duration) // 166 years
 *    println(_duration to TimeUnit.DAY) // 60590 days
 * ```
 *
 * If a duration is converted to bigger unit, it will be truncated.
 *
 *
 * @property value value of the duration.
 * @property unit unit of the duration
 * @constructor Creates a new duration
 * @see TimeUnit
 */
class Duration(val value: Long, val unit: TimeUnit) {
	
	// region Properties
	/**
	 *  `true` if the duration is zero
	 */
	fun isZero(): Boolean = value == 0L
	
	/**
	 *  `true` if the duration is not zero
	 */
	fun isNotZero(): Boolean = value != 0L
	
	/**
	 * Millisecond equivalent of the current duration
	 */
	val asMilliseconds: Long by ValueConvertor(TimeUnit.MILLISECOND)
	
	/**
	 *  Seconds equivalent of the current duration
	 */
	val asSeconds: Long by ValueConvertor(TimeUnit.SECOND)
	
	/**
	 *  Minutes equivalent of the current duration
	 */
	val asMinutes: Long by ValueConvertor(TimeUnit.MINUTE)
	
	/**
	 *  Hours equivalent of the current duration
	 */
	val asHours: Long by ValueConvertor(TimeUnit.HOUR)
	
	/**
	 *  Days equivalent of the current duration
	 */
	val asDays: Long by ValueConvertor(TimeUnit.DAY)
	
	/**
	 *  Months equivalent of the current duration
	 */
	val asMonths: Long by ValueConvertor(TimeUnit.MONTH)
	
	/**
	 *  Years equivalent of the current duration
	 */
	val asYears: Long by ValueConvertor(TimeUnit.YEAR)
	
	// endregion
	
	// region private class ValueConvertor(private val unit: TimeUnit) {
	private class ValueConvertor(private val unit: TimeUnit) {
		@Suppress("NOTHING_TO_INLINE")
		inline operator fun getValue(ref: Duration, property: KProperty<*>): Long = when (unit) {
			
			TimeUnit.MILLISECOND -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value
				TimeUnit.SECOND      -> ref.value * 1000
				TimeUnit.MINUTE      -> ref.value * 1000 * 60
				TimeUnit.HOUR        -> ref.value * 1000 * 60 * 60
				TimeUnit.DAY         -> ref.value * 1000 * 60 * 60 * 24
				TimeUnit.MONTH       -> ref.value * 1000 * 60 * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 1000 * 60 * 60 * 24 * 365
			}
			
			TimeUnit.SECOND      -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 1000
				TimeUnit.SECOND      -> ref.value
				TimeUnit.MINUTE      -> ref.value * 60
				TimeUnit.HOUR        -> ref.value * 60 * 60
				TimeUnit.DAY         -> ref.value * 60 * 60 * 24
				TimeUnit.MONTH       -> ref.value * 60 * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 60 * 60 * 24 * 365
			}
			
			TimeUnit.MINUTE      -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 60000
				TimeUnit.SECOND      -> ref.value / 60
				TimeUnit.MINUTE      -> ref.value
				TimeUnit.HOUR        -> ref.value * 60
				TimeUnit.DAY         -> ref.value * 60 * 24
				TimeUnit.MONTH       -> ref.value * 60 * 24 * 30
				TimeUnit.YEAR        -> ref.value * 60 * 24 * 365
			}
			
			TimeUnit.HOUR        -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 3600000
				TimeUnit.SECOND      -> ref.value / 3600
				TimeUnit.MINUTE      -> ref.value / 60
				TimeUnit.HOUR        -> ref.value
				TimeUnit.DAY         -> ref.value * 24
				TimeUnit.MONTH       -> ref.value * 24 * 30
				TimeUnit.YEAR        -> ref.value * 24 * 365
			}
			
			TimeUnit.DAY         -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 86400000
				TimeUnit.SECOND      -> ref.value / 86400
				TimeUnit.MINUTE      -> ref.value / 1440
				TimeUnit.HOUR        -> ref.value / 24
				TimeUnit.DAY         -> ref.value
				TimeUnit.MONTH       -> ref.value * 30
				TimeUnit.YEAR        -> ref.value * 365
			}
			
			TimeUnit.MONTH       -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 2592000000
				TimeUnit.SECOND      -> ref.value / 2592000
				TimeUnit.MINUTE      -> ref.value / 43200
				TimeUnit.HOUR        -> ref.value / 720
				TimeUnit.DAY         -> ref.value / 30
				TimeUnit.MONTH       -> ref.value
				TimeUnit.YEAR        -> ref.value * 12
			}
			
			TimeUnit.YEAR        -> when (ref.unit) {
				TimeUnit.MILLISECOND -> ref.value / 31536000000
				TimeUnit.SECOND      -> ref.value / 31536000
				TimeUnit.MINUTE      -> ref.value / 525600
				TimeUnit.HOUR        -> ref.value / 8760
				TimeUnit.DAY         -> ref.value / 365
				TimeUnit.MONTH       -> ref.value / 12
				TimeUnit.YEAR        -> ref.value
			}
		}
	}
	// endregion
	
	// region private class UnitConvertor(private val unit: TimeUnit) {
	private class UnitConvertor(val unit: TimeUnit) {
		@Suppress("NOTHING_TO_INLINE")
		inline operator fun getValue(ref: Duration, property: KProperty<*>): Duration = ref.asUnit(unit)
	}
	// endregion
	
	// region Member functions
	fun asUnit(unit: TimeUnit): Duration {
		return when (unit) {
			TimeUnit.MILLISECOND -> Duration milliseconds asMilliseconds
			TimeUnit.SECOND      -> Duration seconds asSeconds
			TimeUnit.MINUTE      -> Duration minutes asMinutes
			TimeUnit.HOUR        -> Duration hours asHours
			TimeUnit.DAY         -> Duration days asDays
			TimeUnit.MONTH       -> Duration months asMonths
			TimeUnit.YEAR        -> Duration years asYears
		}
	}
	
	infix fun to(unit: TimeUnit): Duration = asUnit(unit)
	fun newDuration(value: Long) = Duration(value, unit)
	// endregion
	
	// region Operator functions
	operator fun plus(duration: Long): Duration = newDuration(value + duration)
	
	operator fun plus(duration: Duration): Duration = newDuration(value + duration.asUnit(unit).value)
	
	operator fun minus(duration: Duration): Duration = newDuration(value - duration.asUnit(unit).value)
	
	operator fun minus(duration: Long): Duration = newDuration(value - duration)
	// endregion
	
	// region Overrides functions 
	override fun toString() = "$value $unit"
	override fun equals(other: Any?): Boolean = other is Duration && value == other.value && unit == other.unit
	override fun hashCode() = Objects.hash(value, unit)
	// endregion
	
	// region Companion functions
	companion object {
		fun newDuration(value: Long, unit: TimeUnit) = Duration(value, unit)
		infix fun milliseconds(value: Long): Duration = Duration(value, TimeUnit.MILLISECOND)
		infix fun seconds(value: Long): Duration = Duration(value, TimeUnit.SECOND)
		infix fun minutes(value: Long): Duration = Duration(value, TimeUnit.MINUTE)
		infix fun hours(value: Long): Duration = Duration(value, TimeUnit.HOUR)
		infix fun days(value: Long): Duration = Duration(value, TimeUnit.DAY)
		infix fun months(value: Long): Duration = Duration(value, TimeUnit.MONTH)
		infix fun years(value: Long): Duration = Duration(value, TimeUnit.YEAR)
	}
	// endregion
}

fun main() {
	
	val _duration = Duration months 2000 to TimeUnit.YEAR
	println(_duration)
	println(_duration to TimeUnit.DAY)
	
}




































